File: MkoForms/FormMetadata.cs
using MkoForms.ControlMetadata;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MkoForms
{
    public class FormMetadata
    {
        public FormMetadata()
        {
            this.controls = new Dictionary<string, GeneralControlMetadata>();
        }

        public Dictionary<string, GeneralControlMetadata> controls;
        public string saveUrl;
        public string okUrl;
        public string cancelUrl;
    }
}

File: AngularForms/src/app/formMetadata.ts
import { NumberControlComponent } from './number-control/number-control.component';
import { GeneralControlMetadata } from './generalControl/generalControlMetadata';
import { TextInputControlBaseMetadata } from './textInputControlBase/textInputControlBaseMetadata';
import { DecimalControlMetadata } from './decimal-control/decimalControlMetadata';
import { IntegerControlMetadata } from './integer-control/integerControlMetadata';
import { StringControlMetadata } from './string-component/stringControlMetadata';
import { CheckboxControlMetadata } from './checkbox-control/checkboxControlMetadata';

export type AnyControlMetadata = GeneralControlMetadata | TextInputControlBaseMetadata | DecimalControlMetadata | IntegerControlMetadata | 
                                 StringControlMetadata | CheckboxControlMetadata | FormGroupMetadata | FormArrayMetadata;
export type ControlsMetadata = { [name: string]: AnyControlMetadata }

export interface FormMetadata {
    controls: ControlsMetadata;
    saveUrl: string;    
    okUrl?: string;
    cancelUrl?: string;
}

export interface FormGroupMetadata {
    controls: ControlsMetadata;
}


export interface FormArrayMetadata {
    minLength?: number;
    maxLength?: number;
    itemMetadata: AnyControlMetadata;
}